kvserver: rationalize timestamps in proposals

andreimatei opened the pull request:
see individual commits

andreimatei committed:
kvserver: more Raft closed timestamp assertions

In #62655 we see that there appears to be something wrong with the Raft
closed timestamp. That issue shows an assertion failure about a command
trying to write below a timestamp that was promised to have been closed
by a previous command. This patch includes a little bit more info in
that assertion (the current lease) and adds another two assertions:
- an assertion that the closed timestamp itself does not regress. This
assertion already existed in stageTrivialReplicatedEvalResult(), but
that comes after the failure we've seen. The point of copying this
assertion up is to ascertain whether we're screwing up by regressing the
closed timestamp, or whether a particular request/command is at fault
for not obeying the closed timestamp.
- an assertion against closed ts regression when copying the replica
state from a staging batch to the replica.

All these assertions can be disabled with an env var if things go bad.

Fixes #62765

Release note: None

andreimatei committed:
kvserverpb: update comment on RER.Timestamp

It was lying about helping with lease protection.

Release note: None

andreimatei committed:
kvserver: switch meaning of RER.Timestamp

The ReplicatedEvalResult.Timestamp used to carry the read timestamp of
the request that proposed the respective command. This patch renames it
to WriteTimestamp and sets it accordingly.

The field is used for two purposes: for bumping the clock on followers
such that they don't have values above their clock, and for checking
that we're not writing below the GC threshold. Both of these use cases
actually want the write timestamps, so they were broken. The second use
seems dubious to me (I don't think it's actually needed) but this patch
does not take a position on it beyond adding a TODO.

Beyond fixing the existing uses of this field, putting the write
timestamp in every Raft command has the added benefit that we can use it
to assert below Raft that nobody tries to write below the closed
timestamp. Before this patch, only the leaseholder was able to do this
assertion (because it was the only replica that had access to the write
timestamp) and so the leaseholder had to do it as a non-deterministic
error (and non-deterministic errors are nasty, even when they're
assertion failures). This patch will help turn the assertion into a
deterministic one in the future.

In addition, by clarifying the meaning of this field, this patch opens
the door for #62569 to muddy the field back and give it a special
meaning for lease transfers - the lease start time.

Also mentioning #55293 because this patch lightly touches on the
interaction between requests and the GC threshold.

Release note: None

andreimatei committed:
kvserver: future-proof RER.WriteTimestamp

This patch makes below-Raft code not freak out if the field is not set
(i.e. it appears to be an empty timestamp). This is in order to leave
the door open for future versions to not send the field any more, or at
least send it only on some commands. Concretely, this patch allows the
proposer to declare that particular request does not need to be checked
against the GC threshold below Raft (whether *any* request actually
needs to be checked below Raft is a question that we leave open).

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63045)
<!-- Reviewable:end -->

andreimatei commented:
@nvanbenschoten I'm curious if you would have done something differently in the last commit other than reading the clock for `!isIntentWrite` requests.

bdarnell reviewed (commented):
> Reviewed 9 of 9 files at r2, 2 of 2 files at r3, 5 of 5 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 464 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXYAz6D8y8YS5eUuBal:-MXYAz6D8y8YS5eUuBam:bi9hu9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_application_state_machine.go#L464)):*
> ```Go
> 
> 		wts := cmd.raftCmd.ReplicatedEvalResult.WriteTimestamp
> 		if !wts.IsEmpty() && wts.LessEq(b.state.RaftClosedTimestamp) {
> ```

The preceding comments are now out of date.

We no longer have any kind of assertion on un-migrated ranges; is that a problem for the transition?

Using the existence of RaftClosedTimestamp as a proxy for "migrated to 21.1" is subtle; can we add a method with a name that indicates this more clearly?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXYAiJw4GnlXdKFmhRw:-MXYAiJw4GnlXdKFmhRx:b-ua1tr8) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_application_state_machine.go#L498)):*
> ```Go
> 		b.maxTS.Forward(clockTS)
> 	}
> 	clockTS := cmd.replicatedResult().ClockSignal.UnsafeToClockTimestamp()
> ```

This method is documented as test-only. Shouldn't this use the `TryTo` version as above? (I'm not familiar with these changes to the hlc package)
___
*[pkg/kv/kvserver/replica_proposal.go, line 811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9Ll33Q4xZAixzcQ5:-MXY9Ll33Q4xZAixzcQ6:by51fe7) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L811)):*
> ```Go
> 		// side-effects that are to be replicated to all replicas.
> 		res.Replicated.IsLeaseRequest = ba.IsLeaseRequest()
> 		// Set the timestamp fields. If this range has not "migrated" to 21.1 (as
> ```

This means that you won't be able to do rolling upgrades from the previous 21.1 beta to this one. That's not great and it's pretty late in the cycle to be introducing a version barrier like this, although it's hard enough to do range-level migrations that we don't have another practical option if we want to get this in to 21.1. 
___
*[pkg/kv/kvserver/replica_proposal.go, line 821 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9g0lAgmqR6OWtijv:-MXY9g0lAgmqR6OWtijw:b-r7uxfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L821)):*
> ```Go
> 			res.Replicated.WriteTimestamp = ba.WriteTimestamp()
> 		} else {
> 			res.Replicated.ClockSignal = r.store.Clock().Now()
> ```

What takes the else branch here? For lease transfers, should we be using the new lease's start time instead of the current time? For things other than lease transfers, do we need to include a clock signal at all? 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9Ll33Q4xZAixzcQ5:-MXYX4Z63t0M_99kl0ar:b-vgykry) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L811)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This means that you won't be able to do rolling upgrades from the previous 21.1 beta to this one. That's not great and it's pretty late in the cycle to be introducing a version barrier like this, although it's hard enough to do range-level migrations that we don't have another practical option if we want to get this in to 21.1. 
</blockquote></details>

Yeah... I think requiring stop-the-world for upgrades away from a beta is not the worst thing in the world, as long as the consequences of not doing that are not the worst. And in this case I think they, unfortunately, would be the worst - if a new node proposes a command without a WriteTimestamp to a combination of new and old nodes, the old ones will reject it and the new ones will apply it -> divergence.
I'll think what to do about it.


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvserver: rationalize proposal timestamp

This patch reworks the ReplicatedEvalResult.WriteTimestamp field
(*).  Before this patch, WriteTimestamp was always coming from
ba.WriteTimestamp(), which is either a transaction's write timestamp or,
for non-txn requests, the batch's read timestamp or, for non-MVCC
requests, some random clock value. Below Raft, the field is used for
updating the followers' clocks, and also to check the request against
the GC threshold. This patch sets the WriteTimestamp differently for
IntentWrite requests than other requests:
- for regular writes, the field remains ba.WriteTimestamp()
- for other proposals, the field is a clock reading on the proposer

Some requests (e.g. LeaseTransfers) need a clock signal to travel with
their proposal, and now they get it (see #62569).

[*] An alternative to split the field into two was considered, but it's
hard to do now because of backwards compatibility. It can be done in the
next release, though, because now all the uses of the WriteTimestamp
field tolerate it being empty.

Fixes #62569

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 464 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXYAz6D8y8YS5eUuBal:-MXZEFeWCWO1gy4JOfdh:b-p00mzo) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_application_state_machine.go#L464)):*
> The preceding comments are now out of date.

done

> We no longer have any kind of assertion on un-migrated ranges; is that a problem for the transition?

We never really had the assertion on un-migrated ranges (it was never firing) because `RaftClosedTimestamp` is zero.

> Using the existence of RaftClosedTimestamp as a proxy for "migrated to 21.1" is subtle; can we add a method with a name that indicates this more clearly?

done
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 498 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXYAiJw4GnlXdKFmhRw:-MXZDr65E3UxUOCjOj6e:bb7gzl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_application_state_machine.go#L498)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

This method is documented as test-only. Shouldn't this use the `TryTo` version as above? (I'm not familiar with these changes to the hlc package)
</blockquote></details>

gone
___
*[pkg/kv/kvserver/replica_proposal.go, line 821 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9g0lAgmqR6OWtijv:-MXcd-U87Y9KO-2GBPrC:b-1it6xj) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L821)):*
> What takes the else branch here? 

Proposals corresponding to single `EndTxns`, inline `DeleteRangeRequest`, `ClearRange`, `RevertRange`, `HeartbeatTxn`, `GC`, `PushTxn`, `RecoverTxn`, `RecomputeStats`, `Migrate`.

> For lease transfers, should we be using the new lease's start time instead of the current time?

Would that be better? From a clock sync perspective, it would be slightly worse because we're missing the opportunity to sync the clocks better. Are you thinking of code clarity or of the cost of reading the clock?

> For things other than lease transfers, do we need to include a clock signal at all?

Propagating a clock signal never hurts... Other than the cost of reading the clock?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 811 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9Ll33Q4xZAixzcQ5:-MXceYiw-k1aFlAM-JOn:b4th2i) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L811)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Yeah... I think requiring stop-the-world for upgrades away from a beta is not the worst thing in the world, as long as the consequences of not doing that are not the worst. And in this case I think they, unfortunately, would be the worst - if a new node proposes a command without a WriteTimestamp to a combination of new and old nodes, the old ones will reject it and the new ones will apply it -> divergence.
I'll think what to do about it.
</blockquote></details>

I've went back to a single `WriteTimestamp` field, and used another field to discriminate "regular writes" from other misc requests.


<!-- Sent from Reviewable.io -->

bdarnell reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 821 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9g0lAgmqR6OWtijv:-MXh3sYCAjdibTezVZi3:bhzn6ju) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L821)):*
> Would that be better? From a clock sync perspective, it would be slightly worse because we're missing the opportunity to sync the clocks better. Are you thinking of code clarity or of the cost of reading the clock?

I'm thinking of code clarity - sending Now() is slightly better for clock sync purposes, but then why not send the `ClockSignal=Now()` field unconditionally? If it's dependent on the proposal type then it seems like it should be based on the semantics of the proposal instead of just a conditional inclusion of a current-time clock signal. 

Anyway, I don't feel strongly about this. I'm fine with including the current time signal (even for proposals that write an intent). I don't think it's worth the brainpower to try to limit this to proposal types where it's strictly necessary or to provide the minimum timestamp required for correctness. 


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvserver: even more closed ts assertions

This patch adds historical information to the assertion against closed
timestamp regressions. We've seen that assertion fire in #61981.
The replica now maintains info about what command last bumped the
ClosedTimestamp.

Release note: None

nvb reviewed (commented):
> Reviewed 9 of 9 files at r5, 9 of 9 files at r6, 2 of 2 files at r7, 9 of 9 files at r8, 10 of 10 files at r9, 2 of 2 files at r10, 2 of 2 files at r11.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/kv/kvserver/replica.go, line 579 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5vjLD79ShmvT3SdFB:-MY5vjLD79ShmvT3SdFC:b-f22mnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac5c397b6b7d538b5244da7633f7d6f4e6ccb40/pkg/kv/kvserver/replica.go#L579)):*
> ```Go
> 		tenantID roachpb.TenantID // Set when first initialized, not modified after
> 
> 		// Historical information about the command that set the closed timestamp.
> ```

Is this last commit just for debugging? Or were you intending on landing it?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 332 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5rbhoDu0CtpA2e_CC:-MY5rbhoDu0CtpA2e_CD:b-fgh8o9) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/replica_application_state_machine.go#L332)):*
> ```Go
> 	isIntentWrite := raftCmd.ReplicatedEvalResult.IsIntentWrite()
> 	wts := raftCmd.ReplicatedEvalResult.WriteTimestamp
> 	if isIntentWrite && !wts.IsEmpty() && wts.LessEq(*replicaState.GCThreshold) {
> ```

Are the changes here backward compatable? What would happen in a mixed v20.2/v21.1 cluster if a request has a timestamp before the GC threshold but is not an intent write? Wouldn't a v20.2 replica reject the request but the v21.1 node would apply it?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 459 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5uZK1BvB1aZWhBfmA:-MY5uZK1BvB1aZWhBfmB:bddwoyc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L459)):*
> ```
> assertNoCmdClosedTimestampRegression
> ```

Did you mean to call this twice? Once here, and once in `stageTrivialReplicatedEvalResult`?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1022 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5utUC6gD_oI4Fv2pr:-MY5utUC6gD_oI4Fv2ps:b-9lhh7c) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L1022)):*
> ```Go
> // down assertions.
> func raftClosedTimestampAssertionsEnabled() bool {
> 	return envutil.EnvOrDefaultBool("COCKROACH_RAFT_CLOSEDTS_ASSERTIONS_ENABLED", true)
> ```

Pull this into a static variable so that it is only computed once. There's no need for the function.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1042 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5vRcc0pkOlCplNiSt:-MY5vRcc0pkOlCplNiSu:b7afsro) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L1042)):*
> ```Go
> 		return errors.AssertionFailedf(
> 			"raft closed timestamp regression in cmd: %x; batch state: %s, command: %s, lease: %s, req: %s",
> 			cmd.idKey, existingClosed.String(), newClosed.String(), b.state.Lease, req)
> ```

Have you confirmed that these two assertion errors aren't redacting all of the interesting information? That would be a shame.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 412 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5w-Bc8fQTbGd4WnJw:-MY5w-Bc8fQTbGd4WnJx:b-okf9r0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac5c397b6b7d538b5244da7633f7d6f4e6ccb40/pkg/kv/kvserver/replica_application_state_machine.go#L412)):*
> ```Go
> 
> type closedTimestampSetterInfo struct {
> 	cmd   *replicatedCmd
> ```

Will this play well with the memory recycling we perform in `replicatedCmdBuf`?
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 113 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5rFR3D0as9OMBSfRm:-MY5rFR3D0as9OMBSfRn:b-a4jszt) ([raw file](https://github.com/cockroachdb/cockroach/blob/221cf3c1a874ee6caf0e3cd90b7c651165821c86/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L113)):*
> ```protobuf
> 
> // EvalResultFlags represent the bits in ReplicatedEvalResult.flags.
> enum EvalResultFlags {
> ```

`ReplicatedEvalResultFlags`?
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 137 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5q8pXBS0MqgtlBT-0:-MY5q8pXBS0MqgtlBT-1:b-p3xqeq) ([raw file](https://github.com/cockroachdb/cockroach/blob/221cf3c1a874ee6caf0e3cd90b7c651165821c86/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L137)):*
> ```protobuf
>   // flags indicates various attributes of the command. The interpretation of
>   // the bits is given by EvalResultFlags.
>   int64 flags = 6;
> ```

Should this be typed with `EvalResultFlags`?
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 121 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5tZTG9jrA-lXAXxNi:-MY5tZTG9jrA-lXAXxNj:be7lfke) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L121)):*
> ```protobuf
>   // intents (unless done outside of a txn or the command corresponds to a 1PC
>   // batch). See ba.IsIntentWrite().
>   IntentWrite = 2;
> ```

nit: Do you mind adding the following to ensure that whoever adds the next flag recognizes that these variants are used in a bitfield?
```
// NextFlag = 4;
```
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 146 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5u8LCAzSdYcGnArl5:-MY5u8LCAzSdYcGnArl6:b-fkb6qq) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L146)):*
> ```protobuf
>   // the command against the GC threshold. For all other commands, the timestamp
>   // is to be interpreted simply as a clock signal from the proposer - the field is
>   // also used to bump the followers' clocks.
> ```

~also~


<!-- Sent from Reviewable.io -->

andreimatei committed:
kvserver: add safe formatters for lease and friends

This patch makes some types survive log redacting. In particular, it
ensures that the assertions added in previous patches don't get too
redacted.

Release note: None

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica.go, line 579 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5vjLD79ShmvT3SdFB:-MY5z5HbCidYLUl4p5_5:bngnyy8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac5c397b6b7d538b5244da7633f7d6f4e6ccb40/pkg/kv/kvserver/replica.go#L579)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this last commit just for debugging? Or were you intending on landing it?
</blockquote></details>

Intending to land it... At least until we figure out the failures.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 332 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5rbhoDu0CtpA2e_CC:-MY6NN5h2N2bZK7ebu8I:b-atlodo) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/replica_application_state_machine.go#L332)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Are the changes here backward compatable? What would happen in a mixed v20.2/v21.1 cluster if a request has a timestamp before the GC threshold but is not an intent write? Wouldn't a v20.2 replica reject the request but the v21.1 node would apply it?
</blockquote></details>

like we discussed, I've inverted the flag so that all proposals from 20.2 pass the check here and deterministic behavior is preserved in mixed 20.2/21.1 clusters is kept.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 459 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5uZK1BvB1aZWhBfmA:-MY6cznh8ygrvh9DRt-q:bufhg2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L459)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

> ```
> assertNoCmdClosedTimestampRegression
> ```

Did you mean to call this twice? Once here, and once in `stageTrivialReplicatedEvalResult`?
</blockquote></details>

It was intentional, but you're right - it's silly. I've remove the call in `stageTrivialReplicatedEvalResult`.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1022 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5utUC6gD_oI4Fv2pr:-MY6cQAV7efzFBSyH-0j:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L1022)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Pull this into a static variable so that it is only computed once. There's no need for the function.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1042 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5vRcc0pkOlCplNiSt:-MYHIKRgDygEdbTWVdPW:b-uzjdtx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ce0b3695a91af0bcb0eaca821575ae9847f05fb2/pkg/kv/kvserver/replica_application_state_machine.go#L1042)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Have you confirmed that these two assertion errors aren't redacting all of the interesting information? That would be a shame.
</blockquote></details>

I've appended a commit to make the types involved not redact.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 412 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5w-Bc8fQTbGd4WnJw:-MY6_CxuELIjctFcXLMI:bdjp7sz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4ac5c397b6b7d538b5244da7633f7d6f4e6ccb40/pkg/kv/kvserver/replica_application_state_machine.go#L412)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Will this play well with the memory recycling we perform in `replicatedCmdBuf`?
</blockquote></details>

ah, I didn't know about that. Good catch. See now.
___
*[pkg/kv/kvserver/replica_proposal.go, line 821 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MXY9g0lAgmqR6OWtijv:-MYHJTLNBnqMzE7pUHYX:b-wu2xgo) ([raw file](https://github.com/cockroachdb/cockroach/blob/aae1695faedb19957d35b49eaa9819797ee678f5/pkg/kv/kvserver/replica_proposal.go#L821)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

> Would that be better? From a clock sync perspective, it would be slightly worse because we're missing the opportunity to sync the clocks better. Are you thinking of code clarity or of the cost of reading the clock?

I'm thinking of code clarity - sending Now() is slightly better for clock sync purposes, but then why not send the `ClockSignal=Now()` field unconditionally? If it's dependent on the proposal type then it seems like it should be based on the semantics of the proposal instead of just a conditional inclusion of a current-time clock signal. 

Anyway, I don't feel strongly about this. I'm fine with including the current time signal (even for proposals that write an intent). I don't think it's worth the brainpower to try to limit this to proposal types where it's strictly necessary or to provide the minimum timestamp required for correctness. 
</blockquote></details>

In the meantime, I've moved back to a single timestamp field. It has different meanings for different requests... But I don't think it's worth introducing a new field.
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 113 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5rFR3D0as9OMBSfRm:-MY6_neY3eJ_U1kTHapm:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/221cf3c1a874ee6caf0e3cd90b7c651165821c86/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L113)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`ReplicatedEvalResultFlags`?
</blockquote></details>

done
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 137 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5q8pXBS0MqgtlBT-0:-MY6a-nM0c6XU18mAJyF:b-1mk1jk) ([raw file](https://github.com/cockroachdb/cockroach/blob/221cf3c1a874ee6caf0e3cd90b7c651165821c86/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L137)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Should this be typed with `EvalResultFlags`?
</blockquote></details>

I don't think so... Cause this is a bit map. Should it?
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 121 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5tZTG9jrA-lXAXxNi:-MY6adzu1jZJ_-ZGELPP:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L121)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: Do you mind adding the following to ensure that whoever adds the next flag recognizes that these variants are used in a bitfield?
```
// NextFlag = 4;
```
</blockquote></details>

done
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 146 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MY5u8LCAzSdYcGnArl5:-MY6aqjV5Z_mpUEiO90p:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/100aa79f3972e280c107a4809e3be4b764530f21/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L146)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

~also~
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_proposal.go, line 849 at r16](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYQmP-86mu4QwBhcsNP:-MYQmP-86mu4QwBhcsNQ:b7d1r8o) ([raw file](https://github.com/cockroachdb/cockroach/blob/9a86d3c386b533e4d553ee8aeaab4b117ca87f59/pkg/kv/kvserver/replica_proposal.go#L849)):*
> ```Go
> 			// its lease.
> 			res.Replicated.WriteTimestamp = r.store.Clock().Now()
> 			res.Replicated.SetNonMVCC()
> ```

There is a problem here, caught by the version-upgrade test. We assign or not assign the new flag based on `r.mu.state.MigratedToRaftClosedTimestamp()` here; it's possible that "the migration" is in flight at this point, the range is not migrated at this point, but the command will be applied *after* the a preceding command does the migration. In that case, we might not be setting the flag when the application-side would like it to be set, and application fails because it appears to it that an "intent write" command is writing below the closed timestamp. I'm not sure what to do about it yet; I'm thinking we should push the setting of this flag inside the proposal buffer, and do it after the request was sequenced with preceding requests that might be doing the migration. 


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I've backtracked on adding a field to proposals to discriminate `IsIntentWrite` commands from non-mvcc ones; it seems too hard to correctly make below-Raft code use it correctly. It seems that we need two flags in order to be properly backwards-compatible - one telling you about the character of a proposal and one telling you whether you can use the first flag.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-:-MYflu4g1UKgHHHtVwLJ:bnv4xc7)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell, @nvanbenschoten, and @tbg)



<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 16 of 16 files at r17, 9 of 9 files at r18, 1 of 1 files at r19, 4 of 4 files at r20, 8 of 8 files at r21, 8 of 8 files at r22, 6 of 6 files at r23.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1041 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhSaMk4VBowvz3q7Bo:-MYhSaMk4VBowvz3q7Bp:b-nfp5xq) ([raw file](https://github.com/cockroachdb/cockroach/blob/fea9351c049e2ec914ceda721b6861505e315115/pkg/kv/kvserver/replica_application_state_machine.go#L1041)):*
> ```Go
> }
> 
> // Assert that the current command is not writing under the closed timestamp.
> ```

nit: consider moving this definition and call to this method up above the corresponding ones for `assertNoCmdClosedTimestampRegression`, as an additional subtle hint that this check does not take into consideration the new closed timestamp, just the previous one.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 384 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhUBspAYjNN8ruQVOF:-MYhUBspAYjNN8ruQVOG:b-eu70jl) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L384)):*
> ```Go
> 	// stageTrivialReplicatedEvalResult.
> 	state                 kvserverpb.ReplicaState
> 	closedTimestampSetter closedTimestampSetterInfo
> ```

Give this guy a comment as well.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 410 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhU0Kx6UwWNtK2g6hO:-MYhU0Kx6UwWNtK2g6hP:b-3fofqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L410)):*
> ```Go
> }
> 
> // closedTimestampSetterInfo contains information about the command that last
> ```

Let's move this down to the bottom of this file.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 417 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhTScF8AHBy0_N-q0f:-MYhTScF8AHBy0_N-q0g:bhbc9xy) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L417)):*
> ```Go
> 	// NOTE(andrei): We're going to hold on to the request after its application
> 	// is complete, which is illegal. But cloning it would be too expensive. We're
> 	// onlt using the save requests in case of assertions firing, so let's take
> ```

"only" "saved"
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 419 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhUI-NA0GkcDxeTY0R:-MYhUI-O3lVrncJUF0AQ:b-jq7m4i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L419)):*
> ```Go
> 	// onlt using the save requests in case of assertions firing, so let's take
> 	// the risk.
> 	req *roachpb.BatchRequest
> ```

I'm a little concerned that we're now going to leak the last BatchRequest run on each Range. On a node with 10k ranges and batches of 10 10KB writes, that's 1GB of leaked memory.

How much do we get out of recording the entire BatchRequest? Maybe we just record the lease, the lease index, and whether the batch was a lease request or not? And if it was a lease request, maybe the previous lease?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 845 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhTrXA9h8TdPSQ2pje:-MYhTrXA9h8TdPSQ2pjf:bt72ft8) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L845)):*
> ```Go
> 	if cts := cmd.raftCmd.ClosedTimestamp; cts != nil && !cts.IsEmpty() {
> 		b.state.RaftClosedTimestamp = *cts
> 		b.closedTimestampSetter.leaseIdx = ctpb.LAI(cmd.leaseIndex)
> ```

Let's encapsulate this in a method on `closedTimestampSetterInfo`.
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 130 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhRzp08q6lE7mCisjE:-MYhRzp08q6lE7mCisjF:b2ydmwg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c703552b65b36cc52713fc55d8584a6189c0fd9/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L130)):*
> ```protobuf
>   // The timestamp at which this command is writing. Used to verify the validity
>   // of the command against the GC threshold and to update the followers'
>   // clocks. If the request that produced this command is not an IntentWrite
> ```

If we're using this field for a clock signal again, does the rename to `write_timestamp` make sense? Should we revert that and keep it as `timestamp`?
___
*[pkg/roachpb/data.go, line 1857 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhVzL38xhCroJ0c5tD:-MYhVzL38xhCroJ0c5tE:b3lucst) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dc05cceea254854317a9374a4b88df06c0946a6/pkg/roachpb/data.go#L1857)):*
> ```Go
> // String implements the fmt.Stringer interface.
> func (s LeaseSequence) String() string {
> 	return strconv.FormatInt(int64(s), 10)
> ```

Was this just useless?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, @nvanbenschoten, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1041 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhSaMk4VBowvz3q7Bo:-MYkXOKKE_W5ZXZW7rsj:b-s7fj4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/fea9351c049e2ec914ceda721b6861505e315115/pkg/kv/kvserver/replica_application_state_machine.go#L1041)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

nit: consider moving this definition and call to this method up above the corresponding ones for `assertNoCmdClosedTimestampRegression`, as an additional subtle hint that this check does not take into consideration the new closed timestamp, just the previous one.
</blockquote></details>

I've moved up the definition, but the call order was intentional. If a command is both a regression and a write below the previous closed ts, I want the regression to take priority (since that probably explains the low write too).
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 384 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhUBspAYjNN8ruQVOF:-MYkGVcJBObxUezUP3sG:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L384)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Give this guy a comment as well.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 410 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhU0Kx6UwWNtK2g6hO:-MYkGbS52mIxUgm6BCJA:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L410)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's move this down to the bottom of this file.
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 417 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhTScF8AHBy0_N-q0f:-MYkGf6lAeYbXHNxN0FT:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L417)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

"only" "saved"
</blockquote></details>

done
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 419 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhUI-NA0GkcDxeTY0R:-MYkMECe1WMwpYd3uoLp:b7o9tq3) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L419)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I'm a little concerned that we're now going to leak the last BatchRequest run on each Range. On a node with 10k ranges and batches of 10 10KB writes, that's 1GB of leaked memory.

How much do we get out of recording the entire BatchRequest? Maybe we just record the lease, the lease index, and whether the batch was a lease request or not? And if it was a lease request, maybe the previous lease?
</blockquote></details>

see now; I'm only saving lease requests, plus a flag for splits and merges. Will separately look into including the raw raft log in these assertion messages.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 845 at r22](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhTrXA9h8TdPSQ2pje:-MYkVcBYEyHO4Mf0n5Jj:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/d531f5f416c56372968cefb23dfc8dff28ed5578/pkg/kv/kvserver/replica_application_state_machine.go#L845)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Let's encapsulate this in a method on `closedTimestampSetterInfo`.
</blockquote></details>

done
___
*[pkg/kv/kvserver/kvserverpb/proposer_kv.proto, line 130 at r20](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhRzp08q6lE7mCisjE:-MYkWD-u8ewmhhZ2AETU:b351rxb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5c703552b65b36cc52713fc55d8584a6189c0fd9/pkg/kv/kvserver/kvserverpb/proposer_kv.proto#L130)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

If we're using this field for a clock signal again, does the rename to `write_timestamp` make sense? Should we revert that and keep it as `timestamp`?
</blockquote></details>

I prefer `write_timestamp`; it's better for "usual" requests, and just as good IMO for the others.
___
*[pkg/roachpb/data.go, line 1857 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYhVzL38xhCroJ0c5tD:-MYkVlta-vFadKUI03Pz:b-meyhm0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4dc05cceea254854317a9374a4b88df06c0946a6/pkg/roachpb/data.go#L1857)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Was this just useless?
</blockquote></details>

mostly useless. It made the `LeaseSequence` format properly with "%s" (now a `%s` would result in an ugly message suggesting that you're using the wrong format for an int). Notice the diff below where a `%s` is changed to `%d`


<!-- Sent from Reviewable.io -->

nvb reviewed (commented):
> Reviewed 4 of 9 files at r25, 3 of 4 files at r27, 1 of 2 files at r28, 1 of 2 files at r29, 7 of 7 files at r30.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1358 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYlNLS4E6j9_bkttWrz:-MYlNLS4E6j9_bkttWs-:byups7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/07fd1ca63001c50b6eb8d27b59b66324255561de/pkg/kv/kvserver/replica_application_state_machine.go#L1358)):*
> ```Go
> 	// requests would be too expensive: cloning the request is expensive and also
> 	// requests can be large in memory.
> 	leaseReq *roachpb.RequestLeaseRequest
> ```

Do we need to clear this when the previous request was not a lease request?
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1361 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYlNQAK9h8YzfkUCUls:-MYlNQAK9h8YzfkUCUlt:b-d7k5wk) ([raw file](https://github.com/cockroachdb/cockroach/blob/07fd1ca63001c50b6eb8d27b59b66324255561de/pkg/kv/kvserver/replica_application_state_machine.go#L1361)):*
> ```Go
> 	// split and merge are set if the request was an EndTxn with the respective
> 	// commit trigger set.
> 	split, merge bool
> ```

Same thing here. Should we be clearing these fields at some point?


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)

___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1358 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYlNLS4E6j9_bkttWrz:-MYpoGPN2ejaCJuLTGry:b-ohzngl) ([raw file](https://github.com/cockroachdb/cockroach/blob/07fd1ca63001c50b6eb8d27b59b66324255561de/pkg/kv/kvserver/replica_application_state_machine.go#L1358)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Do we need to clear this when the previous request was not a lease request?
</blockquote></details>

done, good catch. I've added a wipe of the whole struct in `record()`.
___
*[pkg/kv/kvserver/replica_application_state_machine.go, line 1361 at r30](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-MYlNQAK9h8YzfkUCUls:-MYpoI4u4rpawWx6ct8l:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/07fd1ca63001c50b6eb8d27b59b66324255561de/pkg/kv/kvserver/replica_application_state_machine.go#L1361)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Same thing here. Should we be clearing these fields at some point?
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 4 of 9 files at r32, 3 of 4 files at r34, 1 of 2 files at r35, 1 of 2 files at r36, 7 of 7 files at r37.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-:-MYq-e7J218oZFgR2pXO:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-:-MYq9xTZ70eA7sigBucz:b-ydi9ay)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> curl: (7) Failed to connect to register.cockroachdb.com port 443: Connection timed out

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63045#-:-MYqtcpX0QYHvE5Gixw9:b-husnih)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andreimatei, @bdarnell, and @tbg)



<!-- Sent from Reviewable.io -->
